### ISSUE TITLE
Getting the below error

### ISSUE THREAD

**Issue by Vishnupk1993** [healenium/healenium-web#155]
**If you didn't see the video how to use Healenium, please, firstly, get it know on link**
https://healenium.io/

**Describe the problem**
*A clear and concise description of what the problem you've faced with*

**Screenshots**
*If applicable, add screenshots to help explain your problem.*

**Please complete the following information:**
1.	Versions of components (web and backend)
2.	Selenium version
3.	Platform (Java, .NET, etc)
4.	Full stacktrace with exception
5.	Logs from healenium-backend container
6. WebDriver creation

**Additional context**
*Add any other context about the problem here. For example, using of additional frameworks like Sizzle library, Robot Framework or JDI, etc.*

**Comment by Vishnupk1993** [comment]
Getting error during execution at the below line , 
seleniumWebdriver = SelfHealingDriver.create(seleniumWebdriver);

**Error**
ChromeDriver was started successfully.
Exception in thread "main" org.openqa.selenium.JavascriptException: javascript error: Cannot read properties of undefined (reading 'toLowerCase')

**Comment by ElenaStepuro** [comment]
Hello @Vishnupk1993 !
Unfortunately, I cannot help because of lack of information.
Please, fill as much fields in issue template as possible to help you.

**Comment by Vishnupk1993** [comment]
Please find more details below.
1) Healenium Plugin version : 1.0.7
2) Healenium Web version : 3.1.6

3) Sample code : 
package healenium;
import java.util.List;
import java.util.concurrent.TimeUnit;
import com.epam.healenium.SelfHealingDriver;
import com.epam.healenium.SelfHealingDriver;
import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.*;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
public class SampleHeal {

    public static void main(String[] args) throws InterruptedException {
        
        WebDriverManager.chromedriver().setup();
        ChromeOptions options = new ChromeOptions();
        WebDriver seleniumWebdriver = new ChromeDriver(options);
        seleniumWebdriver = SelfHealingDriver.create(seleniumWebdriver);
        seleniumWebdriver.manage().timeouts().implicitlyWait(4, TimeUnit.SECONDS);
        seleniumWebdriver.manage().window().maximize();
		seleniumWebdriver.manage().window().setSize(new Dimension(108,240));
        seleniumWebdriver.get("xyz.com");
		seleniumWebdriver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);
		seleniumWebdriver.findElement(By.xpath("abc")).sendKeys("abc");
		seleniumWebdriver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);
		seleniumWebdriver.findElement(By.xpath("abc")).sendKeys("abc");
		seleniumWebdriver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);
		seleniumWebdriver.findElement(By.xpath("abc")).click();
		seleniumWebdriver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);
		Thread.sleep(3000);
        seleniumWebdriver.findElement(By.xpath("abc")).click();
        seleniumWebdriver.manage().timeouts().implicitlyWait(20, TimeUnit.SECONDS);
		Thread.sleep(2000);
		seleniumWebdriver.findElement(By.xpath("abc")).click();
        seleniumWebdriver.manage().timeouts().implicitlyWait(20, TimeUnit.SECONDS);
		Thread.sleep(2000);
        seleniumWebdriver.findElement(By.xpath("abc")).click();
        seleniumWebdriver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);

  }
}

4) Error screenshot :-
![image](https://user-images.githubusercontent.com/93113082/140753739-de57b68d-e11e-4908-9fc3-54b2d62abb92.png)

**Comment by ElenaStepuro** [comment]
Hello @Vishnupk1993 !
Unfortunately, for now healenium not integrated with Selenium 4.0.0. Please, use 3.141.59 version if you can. To know about related updates you can see this issue https://github.com/healenium/healenium/issues/1
Also, please, add another one container healenium/hlm-selector-imitator:1, as in example, [docker-compose](https://github.com/healenium/healenium-example-maven/blob/master/infra/docker-compose.yaml)
In addition, do not forget to add [healenium.properties](https://github.com/healenium/healenium-example-maven/blob/master/healenium-web-selenium/src/test/resources/healenium.properties) file to your project

**Comment by Vishnupk1993** [comment]
Hi @ElenaStepuro !
Thanks for the quick update.

Tried the same. I have changed the version to 3.141.59 version, added another container. But still we are getting the same error.
Attaching the screenshots:-
1)Docker :- 
![image](https://user-images.githubusercontent.com/93113082/140871760-e2eaf522-472b-4f3d-91ea-05326fe6b017.png)
2) Error Screenshot:- 
![image](https://user-images.githubusercontent.com/93113082/140872222-069bc083-0b6a-463d-bcfa-9f29ae26c7ea.png)
3)Selenium version:-
![image](https://user-images.githubusercontent.com/93113082/140872305-5feb139f-254f-4e37-8a46-23c17d8c9f1a.png)

**Comment by ElenaStepuro** [comment]
Hello @Vishnupk1993 !
In this case, it could be because of incompatibility of chrome driver, chrome browser and selfhealingdriver. I recommend to change chrome driver version to use chrome 92 if you can

**Comment by ElenaStepuro** [comment]
In addition, @Vishnupk1993 , please, send full stacktrace of problem and pom.xml to check your dependencies.
I've rechecked your test example on my machine and it run without problems.

**Comment by Vishnupk1993** [comment]
Hi @ElenaStepuro !
Please find the details below :-
POM XML : 
[Pomhealenium.txt](https://github.com/healenium/healenium-web/files/7527983/Pomhealenium.txt)
Logs :- 
[Logs_Healenium.txt](https://github.com/healenium/healenium-web/files/7527998/Logs_Healenium.txt)

Note : - We tried the same in Salesforce community edition. 
My current chrome version is 95, Please let me know how to check in a lower version through selenium.

**Comment by ElenaStepuro** [comment]
Hi @Vishnupk1993 !
I've rechecked your example on Chrome 95. No worries about it: you don't need to decrease browser version. But I'm worried about this warning from your log:
`[1636722946.998][WARNING]: This version of ChromeDriver has not been tested with Chrome version 95.`

I use chromedriver 95.0.4638.69 and your version is ChromeDriver 94.0.4606.61. Try to increase chromedriver version.
As I can see from log file, it's not healenium issue. It's somewhere in Selenium, most possible it's browser and driver integration.

**Comment by ElenaStepuro** [comment]
Hi @Vishnupk1993 !
Interesting... can we have a call?

**Comment by ElenaStepuro** [comment]
Hello @Vishnupk1993 
You can find stored web elements in database using hlm-backend container. And yes, you can view those web elements from db through docker

**Comment by Vishnupk1993** [comment]
Thanks for the update @ElenaStepuro , we tried in salesforce lightning and only the login part is getting healed. But we are not getting the report , is there any other dependency or do we need to add any extra line code to get the same ?

**Comment by ElenaStepuro** [comment]
@Vishnupk1993 in previous comment forgot to say about other properties...
Please, see my comment below:

- recovery-tries=1 - correct and you can change
- score-cap=0.5 - correct and you can change
- heal-enabled=true - correct and you can change
- serverHost = localhost - correct and you can change
- serverPort = 4567 - **set the value 7878**
- imitatePort - **needs to be added with the value 8000**

**Comment by Vishnupk1993** [comment]
@ElenaStepuro  till now the same issue is getting repeated. Since it's my official code i cannot have a session with you too :( . Can you share me the latest updated yml , helaenium ppty files ?

**Comment by Vishnupk1993** [comment]
i'm trying by adding break point's to check if the code is working or not. Please find the below step which i did during execution

Used intellij
I have written a login code only, 

1) so first time i run the code
2) Then next time i will add a break point in the login xpath
3) Once the screen is visible it will stop before clicking the login button
4) There i manually update the login xpath and resume the execution. 

Previously this was working fine for me then i updated the pom and yml i guess

**Comment by ElenaStepuro** [comment]
Hello @Vishnupk1993 
The latest yaml file you can find here: https://github.com/healenium/healenium-example-maven/blob/master/infra/docker-compose.yaml The latest back version here is 3.2.0 and it's compatible only with web 3.2.0 and 3.2.1 versions
The latest healenium properties you can find here: https://github.com/healenium/healenium-example-maven/blob/master/src/test/resources/healenium.properties
In addition, you can see this setup tutorial: https://www.youtube.com/watch?v=Ed5HyfwZhq4

As about your steps, you don't need to update login xpath in code. It should be changed in DOM om page

**Comment by Vishnupk1993** [comment]
HI @ElenaStepuro  

I tried the same in my Personal Laptop and the salesforce script which i have written is working fine, but the report is not generating. If possible can we have a connect to resolve the same ?

Similarly we are also facing some issue with healing driver in some script's too.

**Comment by ElenaStepuro** [comment]
Hello @Vishnupk1993 , yes, if you need we can have a call. Please, send me by e-mail elena_stepuro@epam.com available time slots for you on the next week.

**Comment by Vishnupk1993** [comment]
Hi @ElenaStepuro ,
Thanks, i have shared you a mail please let me know about the time. If you are occupied with the mentioned IST please let me know your available time.

**Comment by Vishnupk1993** [comment]
@ElenaStepuro We are waiting in the lobby for the clarification.

**Comment by ElenaStepuro** [comment]
@Vishnupk1993 I'm also in the lobby)

**Comment by Vishnupk1993** [comment]
i thing you joined in the other lobby.

**Comment by Vishnupk1993** [comment]
Hi @ElenaStepuro ,

Thanks for the clarification's , as discussed I have attached the folder in my drive and shared you. Please let me know if you require any further input's from our side.

**Comment by Vishnupk1993** [comment]
Hi @ElenaStepuro  any update on the same ? Thanks

**Comment by ElenaStepuro** [comment]
Hello @Vishnupk1993 , unfortunately, I didn't see your test example as we've discussed via call. Could you please resend it?

**Comment by ElenaStepuro** [comment]
Hello @Vishnupk1993 could you please, get an access to e-mail elenastepuro5005@gmail.com?

**Comment by Vishnupk1993** [comment]
HI @ElenaStepuro  Hope you got the access to the folder

**Comment by Vishnupk1993** [comment]
95214 @ElenaStepuro  Please use this code

**Comment by Vishnupk1993** [comment]
*952143

**Comment by raghunadh10** [comment]
> Hi @ElenaStepuro ! We did exactly like you said but still the issue exists , Please find the below logs. [Logs_Healenium.txt](https://github.com/healenium/healenium-web/files/7537651/Logs_Healenium.txt)

Hi 

> Hi @ElenaStepuro greetings, any further update on that issue ?



> Hi @ElenaStepuro ! We did exactly like you said but still the issue exists , Please find the below logs. [Logs_Healenium.txt](https://github.com/healenium/healenium-web/files/7537651/Logs_Healenium.txt)

HI @Vishnupk1993  . I have encountered a smiliar issue which is "javascript error: Cannot read properties of undefined (reading 'toLowerCase')" . i have been using the correct versions of healenium web , backend and chrome. Healenium is working on other scenarios but not on this particular scenario where it is trying to  find the web element . instead of finding it , it is throwing this error. Also when the disbale healing mode is on , it is running fine . can you please check out this problem and help me .

**Comment by ElenaStepuro** [comment]
Hi @raghunadh10 unfortunately, for now Healenium doesn't support Js executor. We work on this issue.

**Comment by Aliaksei-Ashukha** [comment]
Hi @raghunadh10 , @Vishnupk1993 , Issue was fixed in hlm-web version 3.3.0

**Comment by Vishnupk1993** [comment]
@Aliaksei-Ashukha  Thanks for the resolution. Can you please share the latest yaml file ?

**Comment by Aliaksei-Ashukha** [comment]
@Vishnupk1993 , Yaml file remain the same. You need to update hlm-web dependency

**Comment by Vishnupk1993** [comment]
HI @Aliaksei-Ashukha  While running the script i met with this below error, can  you please provide a resolution for this ?
**java: warning: source release 10 requires target release 1.10**

**Comment by raghunadh10** [comment]
Hi @Aliaksei-Ashukha  . thanks for the update  . i have one doubt regarding database we are using in healenium . can we have the ability to change the constraints on columns  like (example -- adding unique constraint to locator column) . will this create any issues ?  please help me on this thing . if any extra details needed, please intimate me .

**Comment by Aliaksei-Ashukha** [comment]
Hi @raghunadh10,
If we talk locator column, there is no constraint of this one.
What exactly column do you want to change ?

**Comment by raghunadh10** [comment]
hi @Aliaksei-Ashukha . In the db , locator column getting stored with duplicate values too , i want to avoid it . so i am thinking to add a unique constraint on it .  so on adding it, will it cause any unreasonable behavior by healenium.

**Comment by Vishnupk1993** [comment]
Hi @Aliaksei-Ashukha tried the same thing and works as expected to me. Thanks for this fix. Quickly, can we set up healenium with eclipse? Because we use our own framework that will support the eclipse rather than intellij.

**Comment by Aliaksei-Ashukha** [comment]
@raghunadh10 , Selector table also have URL, CLASS_NAME, METHOD_NAME. These values used for identify unique of locator. In case you have duplicate locator value, so URL, CLASS_NAME, METHOD_NAME should have other other value. (For example locator By.id("button"); can used on different pages.)

**Comment by Aliaksei-Ashukha** [comment]
@Vishnupk1993 , Healenium does not depend on the IDE. You can use eclipse as well.

**Comment by raghunadh10** [comment]
@Aliaksei-Ashukha  , yeah right . i just saw it . thank you so much for the explanation . one last thing ashukha , so no matter how many times locator get stored or method_name got stored, healenium heals the webelement when its xpath got changed without any trouble right . the reason im asking this is for some web elements its unable to heal . its usually saying new element locator not found  . may be im doing any mistake . can you please give some tips to make the best use of this healing ?  thanks for your time again

**Comment by Vishnupk1993** [comment]
Thanks for the update @Aliaksei-Ashukha , is there any plugin available for eclipse ? Like the Healenium plugin we used for intellij ?

**Comment by Vishnupk1993** [comment]
Hi @Aliaksei-Ashukha I tried running the script through maven and for me the browser is not getting opened im getting these exceptions,  can you please help me to debug this ?
[error log healenium.txt](https://github.com/healenium/healenium-web/files/9363741/error.log.healenium.txt)

**Comment by raghunadh10** [comment]
yeah @Vishnupk1993  . for me its opening fine . based on the logs, i think that  mvn test command hadn't been used . Make sure you run the mvn test command to test the scripts .

**Comment by Vishnupk1993** [comment]
Hi @raghunadh10 I tired the same , Mvn Clean ->MVNValidate -> MVN compile ->MVN test

**Comment by raghunadh10** [comment]
@Vishnupk1993  i want to mention few things before running tests in maven  .
**1) Make sure your test class files are under src/test/*  .
2) Make sure your test class file name ends with Test or Tests or TestCase . ( **ex** - healeniumTest.java ) 
3)Make sure you have added the maven sure fire plugin in the pom.xml .** 

please check these before running and if still facing issue , please let me know

**Comment by raghunadh10** [comment]
Hi @Aliaksei-Ashukha  . One constant issue im getting is the elements which are present in the iframe are not getting healed by healenium . can you look into this once? .

**Comment by Aliaksei-Ashukha** [comment]
@raghunadh10, Yes, you're right, healenium tries to heal every time while element not found on a page. 
Kindly create new issue and provide all necessary info to help us to investigate the bug.

**Comment by Aliaksei-Ashukha** [comment]
Hi @Vishnupk1993 , unfortunately no, there is no healenium plugin for Eclipse.

**Comment by Vishnupk1993** [comment]
hi @raghunadh10  was it working for iframe? I saw one of your comment but now deleted. Please confirm and so that we can raise an new issue for the same.

**Comment by raghunadh10** [comment]
Hi @Vishnupk1993  . it isnt working yet vishnu . so i removed here and created new issue for help regarding that . can you also please confirm whether its working for iframe .  Also vishnu ,  i also have trouble running mvn test , when healenium dependency is added ,mvn test command is not picking up any tests , when that dependency is removed , the tests are running fine with mvn test command. do u still face the same issue ?

**Comment by Vishnupk1993** [comment]
Hi @raghunadh10  Please let me know the defect raised for the same. It's not working for me also.  For me other than the frames it's working perfect

**Comment by raghunadh10** [comment]
Hi @Vishnupk1993  i have raised the defect for the mvn test command .  Please check .  if any inputs are also there from your side, please mention in the comments.
https://github.com/healenium/healenium-web/issues/205

**Comment by Vishnupk1993** [comment]
@Aliaksei-Ashukha @ElenaStepuro @raghunadh10  For all the guidance's, i have given an demonstration of healenium implemented with one of our automation script. And worked well. 

But the question raised during the demo is that , is there anyway we can do the backed process with our normal sql . They are not accepting to use Postgress and Docker. So is there any idea for setting up the same through sql or any other ways ?

**Comment by raghunadh10** [comment]
Hi @Aliaksei-Ashukha  . congrats for your demo.  coming to your question, i am not sure about whether we can use sql or not . but we can use healenium without docker . i have implemented it in our part of project without docker only . installed every service on my local machine and tried. it worked well.

**Comment by Aliaksei-Ashukha** [comment]
Hi @Vishnupk1993 , Could you contact with me directly and we can discuss about paid support.
<img width="261" alt="image" src="https://user-images.githubusercontent.com/69298932/188132343-44d56deb-8ee7-4b5f-a870-ec5fe782e8bc.png">

**Comment by raghunadh10** [comment]
Hi @Aliaksei-Ashukha  @ElenaStepuro  .I am from British Telecom organisation.  After my demo of healenium on automation script  to the team in
my project .we are looking to integrate this healenium
in our project completely.For this  our team need info regarding the fix of salesforce error when running with healenium
like ,is it a fixable issue or not ? . if you are okk to connect ,we can connect whenever it is feasible and 
discuss more regarding this issue and the paid support provided by you which you have mentioned earlier for vishnu.
